File: expense-tracker/expense-tracker-frontend/src/services/expenseService.js
import axios from 'axios';

const API_BASE_URL = 'http://localhost:8080/expenses';

const expenseService = {
  // Fetch all expenses
  getAllExpenses: () => {
    return axios.get(API_BASE_URL);
  },

  // Fetch a single expense by ID
  getExpenseById: (id) => {
    return axios.get(`${API_BASE_URL}/${id}`);
  },

  // Create a new expense
  addExpense: (expense) => {
    return axios.post(API_BASE_URL, expense);
  },

  // Update an existing expense by ID
  updateExpense: (id, expense) => {
    return axios.put(`${API_BASE_URL}/${id}`, expense);
  },

  // Delete an expense by ID
  deleteExpense: (id) => {
    return axios.delete(`${API_BASE_URL}/${id}`);
  },

  // Search expenses by title query parameter
  searchExpensesByTitle: (title) => {
    return axios.get(`${API_BASE_URL}/search`, { params: { title } });
  },

  // Filter expenses by category parameter
  filterExpensesByCategory: (category) => {
    return axios.get(`${API_BASE_URL}/category/${category}`);
  }
};

export default expenseService;

File: expense-tracker/src/main/java/com/expensetracker/service/ExpenseService.java
package com.expensetracker.service;

import com.expensetracker.exception.ResourceNotFoundException;
import com.expensetracker.model.Expense;
import com.expensetracker.repository.ExpenseRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class ExpenseService {

    private final ExpenseRepository expenseRepository;

    @Autowired
    public ExpenseService(ExpenseRepository expenseRepository) {
        this.expenseRepository = expenseRepository;
    }

    // Add a new expense
    public Expense addExpense(Expense expense) {
        return expenseRepository.save(expense);
    }

    // Retrieve all expenses
    public List<Expense> getAllExpenses() {
        return expenseRepository.findAll();
    }

    // Retrieve a single expense by ID
    public Expense getExpenseById(String id) {
        return expenseRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Expense not found with id: " + id));
    }

    // Update an existing expense
    public Expense updateExpense(String id, Expense expenseDetails) {
        Expense expense = getExpenseById(id);
        
        expense.setTitle(expenseDetails.getTitle());
        expense.setAmount(expenseDetails.getAmount());
        expense.setCategory(expenseDetails.getCategory());
        expense.setPaymentMethod(expenseDetails.getPaymentMethod());
        expense.setDate(expenseDetails.getDate());
        expense.setDescription(expenseDetails.getDescription());
        
        return expenseRepository.save(expense);
    }

    // Delete an expense by ID
    public void deleteExpense(String id) {
        Expense expense = getExpenseById(id);
        expenseRepository.delete(expense);
    }

    // Search expenses by title
    public List<Expense> searchExpensesByTitle(String title) {
        return expenseRepository.findByTitleContainingIgnoreCase(title);
    }

    // Filter expenses by category
    public List<Expense> filterExpensesByCategory(String category) {
        return expenseRepository.findByCategoryIgnoreCase(category);
    }

    // Filter expenses by date range
    public List<Expense> filterExpensesByDate(java.time.LocalDate startDate, java.time.LocalDate endDate) {
        return expenseRepository.findByDateBetween(startDate, endDate);
    }
}
